src/heifimage.cpp

```diff
@@ -0,0 +1,235 @@
+#ifdef EXIV2_DEBUG_MESSAGES
+                    std::cerr << "Exiv2::HeifImage::readMetadata: " << err.message << std::endl;
+#endif
+                    throw Error(kerFailedToReadImageData);
+                }
+
+		// hexdump (std::cerr, data, data_size);
+
+		ByteOrder bo = ExifParser::decode(exifData_, data + 10, data_size - 10);
+                setByteOrder(bo);
+                if (data_size > 0 && byteOrder() == invalidByteOrder)
+		{
+#ifndef SUPPRESS_WARNINGS
+                    EXV_WARNING << "Failed to decode Exif metadata.\n";
+#endif
+                    exifData_.clear();
+                }
+	    }
+            heif_image_handle_release(handle);
+        }
+
+    } // HeifImage::readMetadata
+
+    void HeifImage::printStructure(std::ostream& out, PrintStructureOption option, int depth)
+    {
+        if (io_->open() != 0)
+            throw Error(kerDataSourceOpenFailed, io_->path(), strError());
+
+        // Ensure that this is the correct image type
+        if (!isHeifType(*io_, false)) {
+            if (io_->error() || io_->eof())
+                throw Error(kerFailedToReadImageData);
+            throw Error(kerNotExpectedFormat);
+        }
+    }  // HeifImage::printStructure
+
+    void HeifImage::writeMetadata()
+    {
+        if (io_->open() != 0)
+        {
+            throw Error(kerDataSourceOpenFailed, io_->path(), strError());
+        }
+        IoCloser closer(*io_);
+        BasicIo::UniquePtr tempIo(new MemIo);
+        assert (tempIo.get() != 0);
+
+        doWriteMetadata(*tempIo); // may throw
+        io_->close();
+        io_->transfer(*tempIo); // may throw
+
+    } // HeifImage::writeMetadata
+
+    void HeifImage::doWriteMetadata(BasicIo& outIo)
+    {
+        if (!io_->isopen()) throw Error(kerInputDataReadFailed);
+        if (!outIo.isopen()) throw Error(kerImageWriteFailed);
+
+#ifdef EXIV2_DEBUG_MESSAGES
+        std::cout << "Exiv2::HeifImage::doWriteMetadata: Writing HEIF file " << io_->path() << std::endl;
+        std::cout << "Exiv2::HeifImage::doWriteMetadata: tmp file created " << outIo.path() << std::endl;
+#endif
+
+        // Ensure that this is the correct image type
+        if (!isHeifType(*io_, true))
+        {
+            if (io_->error() || io_->eof()) throw Error(kerInputDataReadFailed);
+            throw Error(kerNoImageInInputData);
+        }
+
+        // Write HEIF Signature.
+        if (outIo.write(HeifSignature, sizeof(HeifSignature)) != sizeof(HeifSignature)) throw Error(kerImageWriteFailed);
+#ifdef EXIV2_DEBUG_MESSAGES
+        std::cout << "Exiv2::HeifImage::doWriteMetadata: EOF" << std::endl;
+#endif
+
+    } // HeifImage::doWriteMetadata
+
+    Image::UniquePtr newHeifInstance(BasicIo::UniquePtr io, bool create)
+    {
+        Image::UniquePtr image(new HeifImage(std::move(io), create));
+        if (!image->good())
+        {
+            image.reset();
+        }
+        return image;
+    }
+
+    bool isHeifType(BasicIo& iIo, bool advance)
```

Comment:
There are the functions
`enum heif_filetype_result heif_check_filetype(const uint8_t* data, int len);`
and
`enum heif_brand heif_main_brand(const uint8_t* data, int len);`
to check the HEIF file type. This is more robust, since the first 10 bytes may look different. Pass in at least 12 bytes to these functions.

Reply:
Danke.